### ISSUE TITLE
Borderless mode does not scale well on multi-monitor set up with different Windows scaling configured

### ISSUE THREAD

**Issue by UselessToucan** [ppy/osu-framework#1900]
Environment:
Windows 10 computer with 2 displays. The primary display is external 1080p monitor(Windows scale: 100%), the secondary one is Surface Pro 2 display which is 1080p too(Windows scale: 150%).

Steps to reproduce:
1. Launch visual-tests in windowed mode.
2. Move it to the secondary display.
3. Switch to borderless. (using `TestCaseFullscreen`)

Expected results:
The content is scaled correctly. The whole scene is visible.

Actual results:
Only part of the scene is visible. Some elements are out of the screen.
![borderless_different_scale](https://user-images.githubusercontent.com/22450664/45418383-06786e80-b68d-11e8-89cf-42a062618617.jpg)

Additional info:
Fullscreen mode scales ok.
If both displays have the same Windows scale, borderless scales correctly too.
![borderless_same_scale](https://user-images.githubusercontent.com/22450664/45418397-12643080-b68d-11e8-94a7-43708cabb3af.jpg)

**Comment by AtomCrafty** [comment]
https://www.youtube.com/watch?v=FBHfMDiGmjw
A temporary fix (at least for windows) is to mark dotnet.exe as DPI aware in the windows compatibility settings, which will disable any form of automatic scaling.
OpenTK has had some problems with HiDPI support in the past:
opentk/opentk#47
opentk/opentk#746

**Comment by Llaurence** [comment]
I have the same issue on my computer even though I don't have multiple monitors:

![screenshot](https://user-images.githubusercontent.com/40535137/45754068-e90e4c00-bc1a-11e8-900b-9ef258e9fe94.png)

Config:
Windows 7 family edition 64bits SP1
osu!lazer 2018.920.0

Parts of dxdiag output:

```
   User DPI Setting: 120 DPI (125 percent)
 System DPI Setting: 96 DPI (100 percent)
    DWM DPI Scaling: Disabled

          Card name: AMD Radeon HD 7800 Series
          Chip type: AMD Radeon Graphics Processor (0x6818)
       Current Mode: 1920 x 1080 (32 bit) (60Hz)
        Native Mode: 1920 x 1080(p) (60.000Hz)
        Driver Name: aticfx64.dll,aticfx64.dll,aticfx64.dll,aticfx32,aticfx32,aticfx32,atiumd64.dll,atidxx64.dll,atidxx64.dll,atiumdag,atidxx32,atidxx32,atiumdva,atiumd6a.cap,atitmm64.dll
Driver File Version: 24.20.13017.1009 (English)
     Driver Version: 24.20.13017.1009
        DDI Version: 11
       Driver Model: WDDM 1.1
  Driver Attributes: Final Retail
   Driver Date/Size: 9/11/2018 20:33:56, 1927264 bytes
          Vendor ID: 0x1002
          Device ID: 0x6818
          SubSys ID: 0x25541458
        Revision ID: 0x0000
 Driver Strong Name: oem5.inf:ATI.Mfg.NTamd64.6.1.1:ati2mtag_R575B:24.20.13017.1009:pci\ven_1002&dev_6818
     Rank Of Driver: 00E62001
        Video Accel: 
   Deinterlace Caps: n/a
       D3D9 Overlay: Not Supported
            DXVA-HD: Not Supported
       DDraw Status: Enabled
         D3D Status: Enabled
         AGP Status: Enabled
```

**Comment by AtomCrafty** [comment]
Does it display properly if you mark the osu! executable as DPI aware in the Windows compatibility settings?

**Comment by Llaurence** [comment]
Hi,

I ticked "Disable display scaling on high DPI settings" and reran osu!lazer but the borderless mode still doesn't fill the screen.

**Comment by AtomCrafty** [comment]
Hmm, not sure if the two issues have the same cause then...

**Comment by Llaurence** [comment]
Should I try to run this https://github.com/AtomCrafty/list-display-modes/releases/tag/v1.1 as DPI aware to see if there's a change in the native resolution?

**Comment by AtomCrafty** [comment]
Yeah that might be a good idea

**Comment by AtomCrafty** [comment]
Hmm, that's unfortunate... This will probably require a custom per-platform solution since to my knowledge OpenTK does not provide access to DPI settings.

**Comment by AtomCrafty** [comment]
Soo... I spent a couple of hours diving into a possible workaround for Windows, here are the results so far:
Interesting WinApi functions:
- SetProcessDpiAwareness
- GetScaleFactorForMonitor
- GetDpiForMonitor

Unfortunately, GetScaleFactorForMonitor seems to output complete and utter garbage (only possible results are 100%, 140% and 180%), even if the process is marked as dpi aware.
It is possible to calculate the actual display scaling factor by reading dpi information, setting the process to dpi aware, and reading them again. This seems to only work once though, since the process cannot be reverted to the "dpi unaware" state.

A more severe issue might be compatibility: According to msdn, all of the aforementioned methods only exist since windows 8.1, so far I couldn't find any information regarding ui scaling on earlier versions.
(Not sure how reliable these numbers are, the documentation seems to contain conflicting information)

If this turns out to be a dead end, another (admittedly very ugly) solution would be a hidden configuration value to manually tell the framework which scale factor to use.

I wrote a simple tool to just output any dpi-related information, it would be nice if you could run that @Llaurence, so I can see whether the functions exist under windows 7.
https://github.com/AtomCrafty/windows-high-dpi-tests/releases/tag/v1.0

For reference:
https://docs.microsoft.com/de-de/windows/desktop/hidpi/high-dpi-desktop-application-development-on-windows

**Comment by Llaurence** [comment]
It seems it can't find any of those functions...

![image](https://user-images.githubusercontent.com/40535137/45916349-b061b300-be64-11e8-8be2-a540959b4fa1.png)

(see https://github.com/AtomCrafty/windows-high-dpi-tests/issues/1 too the first time I ran your program it crashed)

**Comment by AtomCrafty** [comment]
`sizeFullscreen` doesn't necessarily equal your monitor resolution, so I don't think this is a viable solution.
It defaults to 9999x9999, in your case the game probably tries to create a window with those dimensions, which is capped to the screen size by Windows. This won't work for multi-monitor setups though, since the window would then span multiple screens.

**Comment by smoogipoo** [comment]
I believe we've resolved this issue osu!-side via ppy/osu#2380 . osu!framework likely just needs an app.manifest added. Using the previously discovered methods is not recommended by Microsoft.

**Comment by AtomCrafty** [comment]
Adding a manifest file with dpiAware set to true to both the framework and test assembly doesn't seem to have any effect. And in Llaurence's case it is lazer itself which doesn't scale correctly, even though it already has a manifest.

I suspect that for the visual tests application (and the framework) the app manifest is simply ignored because neither of them are the entry assembly (`dotnet.exe` is).
[As mentioned above](https://github.com/ppy/osu-framework/issues/1900#issuecomment-421837122), changing dpi behavior on `dotnet.exe` yields the result we're looking for.

**Comment by AtomCrafty** [comment]
Is this still relevant? I don't think there is much we can do about it, apart from telling potential users of the framework to add an app manifest to their main executable (which may be dotnet itself, which we don't have control over).

**Comment by Llaurence** [comment]
I don't have the issue on 2018.1017.0!

**Comment by AtomCrafty** [comment]
That's good to hear :)

**Comment by smoogipoo** [comment]
There are two issues in here.

1. The issue with DPI specific to osu-framework. This is the OP, please keep on this topic.
2. The issue by @Llaurence which is specific to osu-lazer and fixed by @AtomCrafty. Please drop this topic.

This is still an issue.

**Comment by Susko3** [comment]
Should be fixed by #4809.
